### PR TITLE
Add oidcng support

### PR DESCRIPTION
**Pull request by MKodde**
**Monolith**
My apologies for this way to big PR. Cutting this feature into smaller parts would greatly increase the development time. I advise reviewing this in a per commit manner. Some, not all, commits have additional information that might assist in the review.

**Installation instructions**
In order to get OpenID Connect TNG to work:
1. Update your OpenConext-deploy instance that is used by your SP Dashboard installation. You should check out the latest changes from `master`
1. Check out the branch from #279
1. In your SP Dashboard directory run: `vagrant provision` and grab a :coffee: 
1. Verify the guest machine's hosts file for the existence of the oidcng hostname. If not add it to the existing 127.0.0.1 entry: `oidcng.dev.support.surfconext.nl`
1. Go back `git checkout feature/oidcng`

You now should have manage & manage-prod on version 4.0.4 with oidc10_rp support enabled. This can be easily checked as you should now see the OIDC tab on the search page.

**Additional notes**
The first two commits implement task 2 of https://www.pivotaltracker.com/story/show/166702113. Be sure to run the migration to be able to test the changes functionally.

Specific web tests for the oidc tng entities have not yet been added. They will be added soon, but this should not prevent the first review session from being performed.

### CODE REVIEW

File: src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Command/Entity/ChooseEntityTypeCommand.php [hunk 1]
```diff
@@ -26,6 +26,11 @@ class ChooseEntityTypeCommand
      */
     private $type;
 
+    /**
+     * @var array
+     */
+    private $choices;
```
Comment (pablothedude) [inline src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Command/Entity/ChooseEntityTypeCommand.php]: Why is this in the command if this could come directly from the ProtocolChoiceFactory?

File: src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Form/Entity/ChooseEntityTypeType.php [hunk 2]
```diff
@@ -27,19 +27,25 @@
...
      * @param FormBuilderInterface $builder
      * @param array $options
      */
     public function buildForm(FormBuilderInterface $builder, array $options)
     {
+        $this->formCount++;
+
+        /** @var ChooseEntityTypeCommand $command */
+        $command = $builder->getData();
+        $choices = $command->getProtocolChoices();
```
Comment (pablothedude) [inline src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Form/Entity/ChooseEntityTypeType.php]: Couldn't the options come from the ProtocolChoiceFactory itself?

File: src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Controller/EntityCreateController.php [hunk 3]
```diff
@@ -38,6 +45,28 @@ class EntityCreateController extends Controller
 {
     use EntityControllerTrait;
 
+    /**
+     * @var ProtocolChoiceFactory
+     */
+    private $protocolChoiceFactory;
+
+    public function __construct(
```
Comment (pablothedude) [inline src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Controller/EntityCreateController.php]: I think it's getting a bit messy now with the EntityControllerTrait and the overloaded constructor.

File: src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Controller/ServiceController.php [hunk 4]
```diff
@@ -212,7 +213,7 @@ public function editAction(Request $request, $serviceId)
             $logger->info(sprintf('Service was edited by: "%s"', '@todo'), (array)$command);
             try {
                 $this->commandBus->handle($command);
-                return $this->redirectToRoute('entity_list', ['serviceId' => $serviceId]);
+                return $this->redirectToRoute('select_service', ['service' => $serviceId]);
```
Comment (pablothedude) [inline src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Controller/ServiceController.php]: Why is this?

File: src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Form/Entity/ChooseEntityTypeType.php [hunk 5]
```diff
@@ -27,19 +27,25 @@
 
 class ChooseEntityTypeType extends AbstractType
 {
+    private $formCount = 0;
```
Comment (pablothedude) [inline src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Form/Entity/ChooseEntityTypeType.php]: Why is this needed?

**Comment by pablothedude** [comment]
After the webtest are added I will do a functional review.

File: app/config/parameters.yml.dist [hunk 6]
```diff
@@ -77,4 +79,7 @@ parameters:
 
     # Playground uri's for OIDC entities
     playground_uri_test: https://test.dev.playground.surfconext.nl
-    playground_uri_prod: https://prod.dev.playground.surfconext.nl
\ No newline at end of file
+    playground_uri_prod: https://prod.dev.playground.surfconext.nl
+    # Playground uri's for OIDC TNG entities
+    oidcng_playground_uri_test: 'https://test.dev.playground.surfconext.nl'
```
[the remark this answers is not in the document]
Comment (MKodde) [reply app/config/parameters.yml.dist]: Yes there are actually, one for each oidc implementation.

File: src/Surfnet/ServiceProviderDashboard/Infrastructure/Manage/Client/QueryClient.php [hunk 7]
```diff
@@ -117,9 +122,17 @@ public function getMetadataXmlByManageId($manageId)
     public function findByManageId($manageId)
     {
         try {
+            // TODO: investigate if we can add the protocol to the param list of this method to prevent the try/retry
```
[the remark this answers is not in the document]
Comment (MKodde) [reply src/Surfnet/ServiceProviderDashboard/Infrastructure/Manage/Client/QueryClient.php]: > I presume this was a notice for yourself during development?

It was initially, but it was not a triviality to add the additional parameter, as this method is used in many different contexts. Making it hard to always include the protocol. I left this todo for future reference and to pick your brain during this review. 

What do you think about this less than elegant solution? Is it acceptable for now or should we invest in a better, more time-consuming solution?

File: src/Surfnet/ServiceProviderDashboard/Application/Command/Entity/DeletePublishedProductionEntityCommand.php [hunk 8]
```diff
@@ -30,12 +30,21 @@ class DeletePublishedProductionEntityCommand implements Command
      */
     private $manageId;
 
+    /**
+     * @var string
+     * @Assert\NotBlank
+     * @Assert\Uuid
+     */
+    private $protocol;
```
[the remark this answers is not in the document]
Comment (MKodde) [reply src/Surfnet/ServiceProviderDashboard/Application/Command/Entity/DeletePublishedProductionEntityCommand.php]: Given the way Manage works, this was the most pragmatic solution. I do feel your feedback is valid. It however will be impossible to figure out the protocol with only the manageId, as the different entity types have different API endpoints.

File: src/Surfnet/ServiceProviderDashboard/Application/Command/Entity/SaveOidcngEntityCommand.php [hunk 9]
```diff
@@ -0,0 +1,1202 @@
...
+     */
+    public function getProtocol()
+    {
+        return $this->protocol;
+    }
+
+    /**
+     * @param string $protocol
+     */
+    public function setProtocol($protocol)
```
[the remark this answers is not in the document]
Comment (MKodde) [reply src/Surfnet/ServiceProviderDashboard/Application/Command/Entity/SaveOidcngEntityCommand.php]: Yes, good point.

File: src/Surfnet/ServiceProviderDashboard/Application/Command/Entity/SaveOidcEntityCommand.php [hunk 10]
```diff
@@ -341,6 +341,11 @@ class SaveOidcEntityCommand implements SaveEntityCommandInterface
      */
     private $manageId;
 
+    /**
+     * @var string
+     */
+    private $protocol;
```
[the remark this answers is not in the document]
Comment (MKodde) [reply src/Surfnet/ServiceProviderDashboard/Application/Command/Entity/SaveOidcEntityCommand.php]: I believe this is a leftover from my first effort where I did not yet make a distinction between the TNG and regular OIDC entity type. I hoped this distinction would only be needed at the API layer.

But you are absolutely right, this is no longer required. Removing this from the command!

File: src/Surfnet/ServiceProviderDashboard/Application/Metadata/JsonGeneratorStrategy.php [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+namespace Surfnet\ServiceProviderDashboard\Application\Metadata;
+
+use Surfnet\ServiceProviderDashboard\Application\Exception\JsonGeneratorStrategyNotFoundException;
+use Surfnet\ServiceProviderDashboard\Domain\Entity\Entity;
+
+/**
+ * Determines which strategy to use when generating Manage json metadata
+ */
+class JsonGeneratorStrategy implements GeneratorInterface
```
[the remark this answers is not in the document]
Comment (MKodde) [reply src/Surfnet/ServiceProviderDashboard/Application/Metadata/JsonGeneratorStrategy.php]: True that!

File: src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/DependencyInjection/Compiler/JsonGeneratorStrategyCompilerPass.php [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+ */
+
+namespace Surfnet\ServiceProviderDashboard\Infrastructure\DashboardBundle\DependencyInjection\Compiler;
+
+use Surfnet\ServiceProviderDashboard\Application\Metadata\JsonGeneratorStrategy;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Reference;
+
+class JsonGeneratorStrategyCompilerPass implements CompilerPassInterface
```
[the remark this answers is not in the document]
Comment (MKodde) [reply src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/DependencyInjection/Compiler/JsonGeneratorStrategyCompilerPass.php]: Personally, I prefer this method. Using tagged services is kind of magic, but a widely adopted Symfony feature. And it fits really well with the strategy pattern.

File: src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Command/Entity/ChooseEntityTypeCommand.php (hunk 1 above)
Comment (MKodde) [reply src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Command/Entity/ChooseEntityTypeCommand.php]: I always find it hard to load dynamic choices on a form. This is my preferred method. Do you suggest a different approach?

File: src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Controller/EntityCreateController.php (hunk 3 above)
Comment (MKodde) [reply src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Controller/EntityCreateController.php]: Yes, I'm with you on this one. We could reconsider the trait solution in a future development effort. For now, I'd like to propose to not change this at this point.

File: src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Controller/ServiceController.php (hunk 4 above)
Comment (MKodde) [reply src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Controller/ServiceController.php]: Good catch, reverting this.

File: src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Form/Entity/ChooseEntityTypeType.php (hunk 5 above)
Comment (MKodde) [reply src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Form/Entity/ChooseEntityTypeType.php]: That enables us to render two forms of the same type on a page. The counter value is used to name the form and thus making it unique.

File: app/config/parameters.yml.dist (hunk 6 above)
[the remark this answers is not in the document]
Comment (pablothedude) [reply app/config/parameters.yml.dist]: But do they differ between oidc and oidcng?

File: src/Surfnet/ServiceProviderDashboard/Infrastructure/Manage/Client/QueryClient.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (pablothedude) [reply src/Surfnet/ServiceProviderDashboard/Infrastructure/Manage/Client/QueryClient.php]: I'm curious what the response from Okke was? For now I think this will be sufficient but if it could be changed with minimal effort then I'll go for that. It's good that this is contained in the infrastructure layer.

File: src/Surfnet/ServiceProviderDashboard/Application/Command/Entity/DeletePublishedProductionEntityCommand.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (pablothedude) [reply src/Surfnet/ServiceProviderDashboard/Application/Command/Entity/DeletePublishedProductionEntityCommand.php]: Then I would still opt to handle this in the infrastructure layer because there it should be and prevents additional logic throughout all layers . You could use the repository to fetch the entity and then delete it based on the instance type. That way only some additional logic will be in the infrastructure layer and won't bubble up trough all the layers.

File: src/Surfnet/ServiceProviderDashboard/Application/Command/Entity/SaveOidcEntityCommand.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (pablothedude) [reply src/Surfnet/ServiceProviderDashboard/Application/Command/Entity/SaveOidcEntityCommand.php]: :+1:

File: src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Command/Entity/ChooseEntityTypeCommand.php (hunk 1 above)
Comment (pablothedude) [reply src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Command/Entity/ChooseEntityTypeCommand.php]: Inject the ProtocolChoiceFactory (or a repository) in the form with the constructor.

File: src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Controller/EntityCreateController.php (hunk 3 above)
Comment (pablothedude) [reply src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Controller/EntityCreateController.php]: I'm fine with that

File: src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Controller/ServiceController.php (hunk 4 above)
Comment (pablothedude) [reply src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Controller/ServiceController.php]: :+1:

File: src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Form/Entity/ChooseEntityTypeType.php (hunk 5 above)
Comment (pablothedude) [reply src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Form/Entity/ChooseEntityTypeType.php]: Yes I where it's for but I don't understand why it's needed.

**Comment by pablothedude** [comment]
Could you add the addition of `oidcng.dev.support.surfconext.nl` to the hosts file to the documentation?

**Comment by pablothedude** [comment]
Unfortunately some additional steps were necessary in order to get the vm and existing oidc entities working.

Functional review:
- The field `Is public client` seems required
- No default value set for `Access token validity`
- 'copytoproduction' changes ClientId (removes protocol)

File: app/config/parameters.yml.dist (hunk 6 above)
[the remark this answers is not in the document]
Comment (MKodde) [reply app/config/parameters.yml.dist]: Yes they will according to @quartje, which seems logical as there is a new OIDC gateway on the block.

File: src/Surfnet/ServiceProviderDashboard/Infrastructure/Manage/Client/QueryClient.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (MKodde) [reply src/Surfnet/ServiceProviderDashboard/Infrastructure/Manage/Client/QueryClient.php]: Due to the way the nosql backend of manage is set up, it is not easy to query over the two protocols. For now that is out of scope. The best thing he could do was to add the protocol to every response from the search API.

We both had a good laugh about the current solution :) And I'm now working on fixing this.

File: src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Command/Entity/ChooseEntityTypeCommand.php (hunk 1 above)
Comment (MKodde) [reply src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Command/Entity/ChooseEntityTypeCommand.php]: Would that work? To my knowledge, Symfony Types do no support dependency injection.

File: src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Form/Entity/ChooseEntityTypeType.php (hunk 5 above)
Comment (MKodde) [reply src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Form/Entity/ChooseEntityTypeType.php]: The protocol selection popup is rendered twice on the page for both test and production entities. And previously they where named the same. This caused problems when handling the form. This was also evident in the front-end form behaviour. For example, clicking on a form label element would focus the corresponding input element on the other, out of sight, form.

**Comment by MKodde** [comment]
> Could you add the addition of `oidcng.dev.support.surfconext.nl` to the hosts file to the documentation?

The changes to the Ansible roles, applied by @quartje should prevent these changes. Where these commits in your branch while running the tests? [1](https://github.com/SURFnet/sp-dashboard/commit/cba2fe86a99eb12f45fb7364395fabc8d12cb42e), [2](https://github.com/SURFnet/sp-dashboard/commit/cba2fe86a99eb12f45fb7364395fabc8d12cb42e) and [3](https://github.com/SURFnet/sp-dashboard/commit/f75fa8ff906503bb17e66ad198ac20388d79f6da). If they where, than we should ask @quartje to review the changes made. 

> And enabling OIDCNG in manage: application.yml:39 `show_oidc_rp: True`

See reply above

File: src/Surfnet/ServiceProviderDashboard/Application/Command/Entity/DeletePublishedProductionEntityCommand.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (MKodde) [reply src/Surfnet/ServiceProviderDashboard/Application/Command/Entity/DeletePublishedProductionEntityCommand.php]: We discussed this IRL. And we are on the same page. For now we're leaving this as is. Added an action to this technical debt chore: https://www.pivotaltracker.com/story/show/167312313 to address this issue at a later stage.

File: src/Surfnet/ServiceProviderDashboard/Domain/Repository/DeleteEntityRepository.php [hunk 13]
```diff
@@ -26,6 +26,7 @@ interface DeleteEntityRepository
      * Delete an entity from manage identified by the manage id
      *
      * @param string $manageId
+     * @param string $protocol
      */
-    public function delete($manageId);
+    public function delete($manageId, $protocol);
```
[the remark this answers is not in the document]
Comment (MKodde) [reply src/Surfnet/ServiceProviderDashboard/Domain/Repository/DeleteEntityRepository.php]: See: https://github.com/SURFnet/sp-dashboard/pull/275#discussion_r304288685

File: src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Command/Entity/ChooseEntityTypeCommand.php (hunk 1 above)
Comment (MKodde) [reply src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Command/Entity/ChooseEntityTypeCommand.php]: Thanks for pointing out IRL that Form types can actually be injected with dependencies when tagged with `form.type`. 

The additional complexion that further prevents moving this logic to the Type, is the fact that we need to know about the active service in order to determine which protocols may be displayed. 
 
This could be fixed by injecting the authz service on the factory. But that would increase coupling and decrease the cohesion..

I'm still opting for this solution.

File: src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Form/Entity/ChooseEntityTypeType.php (hunk 2 above)
Comment (MKodde) [reply src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Form/Entity/ChooseEntityTypeType.php]: See: https://github.com/SURFnet/sp-dashboard/pull/275#discussion_r304302958

File: src/Surfnet/ServiceProviderDashboard/Application/Command/Entity/DeletePublishedProductionEntityCommand.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (pablothedude) [reply src/Surfnet/ServiceProviderDashboard/Application/Command/Entity/DeletePublishedProductionEntityCommand.php]: :+1:
